cmd/frostfs-node/cache.go

```diff
@@ -125,73 +123,60 @@ func (c *lruNetCache) get(key interface{}) (interface{}, error) {
 
 // wrapper over TTL cache of values read from the network
 // that implements container storage.
-type ttlContainerStorage ttlNetCache
+type ttlContainerStorage struct {
+	*ttlNetCache[cid.ID, *container.Container]
+}
 
-func newCachedContainerStorage(v container.Source, ttl time.Duration) *ttlContainerStorage {
+func newCachedContainerStorage(v container.Source, ttl time.Duration) ttlContainerStorage {
 	const containerCacheSize = 100
 
-	lruCnrCache := newNetworkTTLCache(containerCacheSize, ttl, func(key interface{}) (interface{}, error) {
-		var id cid.ID
-
-		err := id.DecodeString(key.(string))
-		if err != nil {
-			return nil, err
-		}
-
+	lruCnrCache := newNetworkTTLCache[cid.ID, *container.Container](containerCacheSize, ttl, func(id cid.ID) (*container.Container, error) {
 		return v.Get(id)
 	})
 
-	return (*ttlContainerStorage)(lruCnrCache)
+	return ttlContainerStorage{lruCnrCache}
 }
 
-func (s *ttlContainerStorage) handleRemoval(cnr cid.ID) {
-	(*ttlNetCache)(s).set(cnr.EncodeToString(), nil, apistatus.ContainerNotFound{})
+func (s ttlContainerStorage) handleRemoval(cnr cid.ID) {
+	s.set(cnr, nil, apistatus.ContainerNotFound{})
 }
 
 // Get returns container value from the cache. If value is missing in the cache
 // or expired, then it returns value from side chain and updates the cache.
-func (s *ttlContainerStorage) Get(cnr cid.ID) (*container.Container, error) {
-	val, err := (*ttlNetCache)(s).get(cnr.EncodeToString())
-	if err != nil {
-		return nil, err
-	}
-
-	return val.(*container.Container), nil
+func (s ttlContainerStorage) Get(cnr cid.ID) (*container.Container, error) {
+	return s.get(cnr)
 }
 
-type ttlEACLStorage ttlNetCache
+type ttlEACLStorage struct {
+	*ttlNetCache[string, *container.EACL]
+}
 
-func newCachedEACLStorage(v container.EACLSource, ttl time.Duration) *ttlEACLStorage {
+func newCachedEACLStorage(v container.EACLSource, ttl time.Duration) ttlEACLStorage {
 	const eaclCacheSize = 100
 
-	lruCnrCache := newNetworkTTLCache(eaclCacheSize, ttl, func(key interface{}) (interface{}, error) {
+	lruCnrCache := newNetworkTTLCache(eaclCacheSize, ttl, func(key string) (*container.EACL, error) {
 		var id cid.ID
 
-		err := id.DecodeString(key.(string))
+		err := id.DecodeString(key)
 		if err != nil {
 			return nil, err
 		}
 
 		return v.GetEACL(id)
 	})
 
-	return (*ttlEACLStorage)(lruCnrCache)
+	return ttlEACLStorage{lruCnrCache}
 }
 
 // GetEACL returns eACL value from the cache. If value is missing in the cache
 // or expired, then it returns value from side chain and updates cache.
-func (s *ttlEACLStorage) GetEACL(cnr cid.ID) (*container.EACL, error) {
-	val, err := (*ttlNetCache)(s).get(cnr.EncodeToString())
-	if err != nil {
-		return nil, err
-	}
-
-	return val.(*container.EACL), nil
+func (s ttlEACLStorage) GetEACL(cnr cid.ID) (*container.EACL, error) {
+	return s.get(cnr.EncodeToString())
```

Comment:
I like how short it gets with generics.

Reply:
Oh, your commment made me think, that we do not need to encode the container ID here, can use `cid.ID` too. Fixed.

Reply:
The only problem I encountered is that you cannot define a type like `type X Type[int, int]` and then some methods on top of it.